cirq-google/cirq_google/workflow/processor_record.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import dataclasses
+
+import cirq
+import cirq_google as cg
+from cirq._compat import dataclass_repr
+from cirq_google.engine.virtual_engine_factory import (
+    MOST_RECENT_TEMPLATES,
+    _create_device_spec_from_template,
+)
+
+
+class ProcessorRecord(metaclass=abc.ABCMeta):
+    """A serializable record that maps to a particular `cg.engine.AbstractProcessor`."""
+
+    @abc.abstractmethod
+    def get_processor(self) -> 'cg.engine.AbstractProcessor':
+        """Using this classes' attributes, return a unique `cg.engine.AbstractProcessor`
+
+        This is the primary method that descendants must implement.
+        """
+
+    def get_sampler(self) -> 'cirq.Sampler':
+        """Return a `cirq.Sampler` for the processor specified by this class.
+
+        The default implementation delegates to `self.get_processor()`.
+        """
+        return self.get_processor().get_sampler()
+
+    def get_device(self) -> 'cirq.Device':
+        """Return a `cirq.Device` for the processor specified by this class.
+
+        The default implementation delegates to `self.get_processor()`.
+        """
+        return self.get_processor().get_device()
+
+
+@dataclasses.dataclass(frozen=True)
+class EngineProcessorRecord(ProcessorRecord):
+    """A serializable record of processor_id to map to a `cg.EngineProcessor`.
+
+    This class presumes the GOOGLE_CLOUD_PROJECT environment
+    variable is set to establish a connection to the cloud service.
+
+    Args:
+        processor_id: The processor id.
+    """
+
+    processor_id: str
+
+    def get_processor(self) -> 'cg.EngineProcessor':
+        """Return a `cg.EngineProcessor` for the specified processor_id."""
+        engine = cg.get_engine()
+        return engine.get_processor(self.processor_id)
+
+    def get_sampler(self) -> 'cg.QuantumEngineSampler':
```

Comment:
Does this need to be reimplemented if the abstract class logic is the same?

Is the more specific `cg.QuantumEngineSampler` type used anywhere over the `cirq.Sampler`?

Reply:
good catch. this was vestigial from when I had to hack in a gateset